tools/tcpbind.py

```diff
@@ -0,0 +1,463 @@
+    u64 cgroupid = bpf_get_current_cgroup_id();
+    if (cgroupset.lookup(&cgroupid) == NULL) {
+      return 0;
+    }""",
+}
+
+# code substitutions
+if args.count:
+    bpf_text = bpf_text.replace("IPV4_CODE", struct_init['ipv4']['count'])
+    bpf_text = bpf_text.replace("IPV6_CODE", struct_init['ipv6']['count'])
+else:
+    bpf_text = bpf_text.replace("IPV4_CODE", struct_init['ipv4']['trace'])
+    bpf_text = bpf_text.replace("IPV6_CODE", struct_init['ipv6']['trace'])
+
+if args.pid:
+    bpf_text = bpf_text.replace('FILTER_PID',
+        'if (pid != %s) { return 0; }' % args.pid)
+if args.port:
+    sports = [int(sport) for sport in args.port.split(',')]
+    sports_if = ' && '.join(['sport != %d' % sport for sport in sports])
+    bpf_text = bpf_text.replace('FILTER_PORT',
+        'if (%s) { currsock.delete(&pid); return 0; }' % sports_if)
+if args.uid:
+    bpf_text = bpf_text.replace('FILTER_UID',
+        'if (uid != %s) { return 0; }' % args.uid)
+if args.errors:
+    bpf_text = bpf_text.replace('FILTER_ERRORS', 'ignore_errors = 0;')
+if args.cgroupmap:
+    bpf_text = bpf_text.replace('FILTER_CGROUP', struct_init['filter_cgroup'])
+    bpf_text = bpf_text.replace(
+        'CGROUP_MAP',
+        (
+            'BPF_TABLE_PINNED("hash", u64, u64, cgroupset, 1024, "%s");' %
+            args.cgroupmap
+        )
+    )
+
+bpf_text = bpf_text.replace('FILTER_PID', '')
+bpf_text = bpf_text.replace('FILTER_PORT', '')
+bpf_text = bpf_text.replace('FILTER_UID', '')
+bpf_text = bpf_text.replace('FILTER_ERRORS', '')
+bpf_text = bpf_text.replace('FILTER_CGROUP', '')
+bpf_text = bpf_text.replace('CGROUP_MAP', '')
+
+# selecting output format - 80 characters or wide, fitting IPv6 addresses
+header_fmt = "%5s %-12.12s %-2s %-15s %-5s %5s %2s"
+output_fmt = b"%5d %-12.12s %-2d %-15.15s %5d %-5s %2d"
+if args.wide:
+    header_fmt = "%10s %-12.12s %-2s %-39s %-5s %5s %2s"
+    output_fmt = b"%10d %-12.12s %-2d %-39s %5d %-5s %2d"
+
+if args.ebpf:
+    print(bpf_text)
+    exit()
+
+if args.debug_source:
+    b = BPF(text=bpf_text, debug=DEBUG_SOURCE)
+    exit()
+
+
+# bind options:
+# SOL_IP     IP_FREEBIND              F....
+# SOL_IP     IP_TRANSPARENT           .T...
+# SOL_IP     IP_BIND_ADDRESS_NO_PORT  ..N..
+# SOL_SOCKET SO_REUSEADDR             ...R.
+# SOL_SOCKET SO_REUSEPORT             ....r
+def opts2str(bitfield):
+    str_options = ""
+    bit = 1
+    for opt in "FTNRr":
+        str_options += opt if bitfield & bit else "."
+        bit *= 2
+    return str_options.encode()
+
+
+# process events
+def print_ipv4_bind_event(cpu, data, size):
+    event = b["ipv4_bind_events"].event(data)
+    global start_ts
+    if args.timestamp:
+        if start_ts == 0:
+            start_ts = event.ts_us
+        printb(b"%-9.6f " % ((float(event.ts_us) - start_ts) / 1000000), nl="")
+    if args.print_uid:
+        printb(b"%6d " % event.uid, nl="")
+    if args.errors:
+        printb(b"%3d " % event.return_code, nl="")
```

Comment:
Maybe we could print non-zero return code with its actual names for better usability? For example, for return_code "-99" prints EADDRNOTAVAIL?

Reply:
It would be hard to satisfy 80-characters limitation. 
I added return code reporting in human-readable format when "wide" output mode is enabled (-w)